README.md
# fluent-plugin-amqp

This plugin provides both a Source and Matcher which uses RabbitMQ as its transport.

[![Build Status](https://travis-ci.org/giraffi/fluent-plugin-amqp.svg?branch=master)](https://travis-ci.org/giraffi/fluent-plugin-amqp)
[![Gem Version](https://badge.fury.io/rb/fluent-plugin-amqp.svg)](https://badge.fury.io/rb/fluent-plugin-amqp)
[![Coverage Status](https://coveralls.io/repos/github/giraffi/fluent-plugin-amqp/badge.svg)](https://coveralls.io/github/giraffi/fluent-plugin-amqp)

# Table of contents

1. [Requirements](#requirements)
1. [Features](#features)
    1. [Highly Available Failover](#feat-failover)
1. [Configuration](#configuration)
    1. [Common parameters](#conf-common)
    1. [Source](#conf-source)
    1. [Matcher](#conf-matcher)
         1. [Message Headers](#conf-matcher-header)
1. [Example Use Cases](#usecases)
    1. [Using AMQP instead of Fluent TCP forwarders](#uc-forwarder)
    1. [Enable TLS Authentication](#uc-tls)
1. [Contributing](#contributing)
1. [Copyright](#copyright)

# Requirements <a name="requirements"></a>


|fluent-amqp-plugin|fluent|ruby|
|----|----|----|
|>= 0.10.0 | >= 0.14.8 | >= 2.1 |
| < 0.10.0 | > 0.10.0, < 2 <sup>*</sup> | >= 1.9  |

* May not support all future fluentd features

# Features <a name="features"></a>

## Highly Available Failover <a name="feat-failover"></a>

You can use the `hosts` parameter to provide an array of rabbitmq hosts which
are in your cluster. This allows for highly avaliable configurations where a
node in your cluster may become inaccessible and this plugin will attempt a reconnection
on another host in the array.

> *WARNING:* Due to limitations in the library being used for connecting to RabbitMQ
each node of the cluster must use the same port, vhost and other configuration.

### Example

```
<source>
  type amqp
  hosts ["amqp01.example.com","amqp02.example.com"]
  port 5672
  vhost /
  user guest
  pass guest
  queue logs
  format json
</source>
```



# Configuration <a name="configuration"></a>

## A note on routing keys

If you would like to filter events from certain sources, you can make use of the
`key`, `tag_key` and `tag_header` configuration options.

The RabbitMQ [routing key](http://www.rabbitmq.com/tutorials/tutorial-four-ruby.html)
that is set for the message on the broker determines what you may be able to
filter against when consuming messages.

For example, if you want a 'catch-all' consumer that gets all messages from a
direct exchange, you should set `tag_key true` on both `source` and `matcher`. This
will then recreate the original event's tag ready for processing by the consumers
matchers.

If you want to have selective control over the messages that are consumed, you
can set `tag_key true` on the matcher, but `key some.tag` on the source. Only
messages with the given tag will be consumed, however its recommended that you
understand the difference between the different exchange types, and how multiple
consumers may impact message delivery.

## Common parameters <a name="conf-common"></a>

The following parameters are common to both matcher and source
plugins, and can be used as required.

|param|type|default|description|
|----|----|----|---|
|:host|:string|nil| *Required (if hosts unset)* Hostname of RabbitMQ server |
|:hosts|:array|nil| *Required (if host unset)* An array of hostnames of RabbitMQ server in a common cluster (takes precidence over `host`)|
|:user|:string|"guest"| Username to connect |
|:pass|:string|"guest"| Password to authenticate with (Secret) |
|:vhost|:string|"/"| RabbitMQ Virtual Host|
|:port|:integer|5672| RabbitMQ listening port|
|:durable|:bool|false| Should the queue or exchange be durable? |
|:passive|:bool|false| If true, will fail if queue or exchange does not exist |
|:auto_delete|:bool|false| Should the queue be deleted when all consumers have closed? |
|:heartbeat|:integer|60| Frequency of heartbeats to ensure quiet connections are kept open|
|:ssl|:bool|false| Is SSL enabled for this connection to RabbitMQ|
|:verify_ssl|:bool|false| Verify the SSL certificate presented by RabbitMQ |
|:tls|:bool|false| Should TLS be used for authentication |
|:tls_cert|:string|nil| Path (or content) of TLS Certificate |
|:tls_key|:string|nil| Path (or content) of TLS Key |
|:tls_ca_certificates|:array|nil| Array of paths to CA certificates |
|:tls_verify_peer|:bool|true| Verify the servers TLS certificate |
|:tag_key|:bool|false| Should the routing key be used for the event tag |
|:tag_header|:string|nil| What header should be used for the event tag |
|:time_header|:string|nil| What header should be used for the events timestamp |


## Source - Obtain events from a RabbitMQ queue <a name="conf-source"></a>


Using the amqp as a source allows you to read messages from RabbitMQ and handle
them in the same manner as a locally generated event.

It can be used in isolation; reading (well formed) events generated by other
applications and published onto a queue, or used with the amqp matcher, which
can replace the use of the fluent forwarders.


### Source specific parameters

Note: The following are in addition to the common parameters shown above.

|param|type|default|description|
|----|----|----|---|
|:tag|:string|"hunter.amqp"| Accepted events are tagged with this string (See also tag_key)|
|:queue|:string|nil| What queue contains the events to read |
|:exclusive|:bool|false| Should we have exclusive use of the queue? See notes on `Multiple Workers` below.|
|:payload_format|:string|"json"| Deprecated - Use `format`|
|:bind_exchange|:boolean|false| Should the queue automatically bind to the exchange |
|:exchange|:string|nil| What exchange should the queue bind to? |
|:exchange_type|:string|"direct"| Type of exchange ( direct, fanout, topic, headers, x-consistent-hash, x-modulus-hash )|
|:routing_key|:string|nil| What exchange should the queue bind to? |
|:include_headers|:bool|false| If true, include Message headers in the parsed payload with key "headers" (if there is already a field name headers in the parsed payload, then it will be overwritten) |

### Example

```
<source>
  type amqp
  host amqp.example.com
  port 5672
  vhost /
  user guest
  pass guest
  queue logs
  format json
</source>
```

## Matcher - output events from RabbitMQ <a name="conf-matcher"></a>

### Matcher specific parameters

|param|type|default|description|
|----|----|----|----|
|:exchange|:string|""| Name of the exchange to send events to |
|:exchange_type|:string|"direct"| Type of exchange ( direct, fanout, topic, headers, x-consistent-hash, x-modulus-hash )|
|:persistent|:bool|false| | Are messages kept on the exchange even if RabbitMQ shuts down |
|:key|:string|nil| Routing key to attach to events (Only applies when `exchange_type topic`) See also `tag_key`|
|:content_type|:string|"application/octet"| Content-type header to send with message |
|:content_encoding|:string|nil| Content-Encoding header to send - eg base64 or rot13 |

#### Headers <a name="conf-matcher-headers"></a>

It is possible to specify message headers based on the content of the incoming
message, or as a fixed default value as shown below;

```
<matcher ...>
...

  <header>
    name LogLevel
    source level
    default "INFO"
  </header>
  <header>
    name SourceHost
    default my.example.com
  </header>
  <header>
    name CorrelationID
    source x-request-id
  </header>
  <header>
    name NestedExample
    source a.nested.value
  </header>
  <header>
    name AnotherNestedExample
    source ["a", "nested", "value"]
  </header>

...
</matcher>
```


The header elements may be set multiple times for multiple additional headers
to be included on any given message.

* If source is omitted, the header will _always_ be set to the default value
* If default is omitted the header will only be set if the source is found
* Overloading headers is permitted
    * Last defined header with a discovered or default value will be used
    * Defaults and discovered values are treated equally - If you set a default
    for a overloaded header the earlier headers *will never be used*


### Example

```
<match **.**>
  type amqp
  key my_routing_key
  exchange amq.direct
  host amqp.example.com
  port 5672
  vhost /
  user guest
  pass guest
  content_type application/json
</match>
```

# Example Use Cases <a name="usecases"></a>

## Using AMQP instead of Fluent TCP forwarders <a name="uc-forwarder"></a>

One particular use case of the AMQP plugin is as an alternative to the built-in
fluent forwarders.

You can simply setup each client to output events to a RabbitMQ exchange
which is then consumed by one or more input agents.

The example configuration below shows how to setup a direct exchange, with
multiple consumers each receiving events.

### Matcher - Writes to Exchange

```
<match **>
  type amqp
  exchange amq.direct
  host amqp.example.com
  port 5672
  vhost /
  user guest
  pass guest
  format json
  tag_key true
</match>
```

### Source - Reads from queues

```
<source>
  type amqp
  host amqp.example.com
  port 5672
  vhost /
  user guest
  pass guest
  queue my_queue
  format json
  tag_key true
</source>
```

## Enable TLS Authentication <a name="uc-tls"></a>

The example below shows how you can configure TLS authentication using signed encryption keys
which will be validated by your appropriately configured RabbitMQ installation.

For more information on setting up TLS encryption, see the [Bunny TLS documentation](http://rubybunny.info/articles/tls.html)

Note: The 'source' configuration accepts the same arguments.

```
<match **.**>
  type amqp
  key my_routing_key
  exchange amq.direct
  host amqp.example.com
  port 5671              # Note that your port may change for TLS auth
  vhost /

  tls true
  tls_key "/etc/fluent/ssl/client.key.pem"
  tls_cert "/etc/fluent/ssl/client.crt.pem"
  tls_ca_certificates ["/etc/fluent/ssl/server.cacrt.pem", "/another/ca/cert.file"]
  tls_verify_peer true
  auth_mechanism EXTERNAL

</match>
```


## Multiple Workers

This plugin supports multiple workers for both source and matcher configurations.

Note that when using exclusive queues with multiple workers the queues will be renamed based on the worker id.

For example, if your queue is configured as `fluent.queue`, with 4 workers and `exclusive: true` the plugin
will create four named queues;

- fluent.queue
- fluent.queue.1
- fluent.queue.2
- fluent.queue.3

Be aware that the first queue will keep the same name as given to maintain compatibility.

## Docker Container

A docker container is included in this project to help with testing and debugging.

You can simply build the docker container's ready for use with the following;
```
docker-compose build
```

Start the cluster of three containers with;
```
docker-compose up
```

And finally, submit test events, one a second, to the built in tcp.socket source
with;

```
while [ true ] ; do echo "{ \"test\": \"$(date)\" }" | nc ${DOCKER_IP} 20001; sleep 1; done
```
## Rabbitmq-sharding

You may find that rabbitmq doesn't behave nicely when delivering lots of events to a single queue as the process thread gets overloaded and starts to send flow control events back to publishers. If you're in this situation, try the rabbitmq-sharding plugin which is in RMQ 3.6+ and can allow queues to be dynamically generated per-node.

To use this;

1. Enable the plugin on all nodes `rabbitmq-plugins enable rabbitmq_sharding`
2. Create an exchange to accept events and to be sharded using `x-modulus-hash` or `x-consistent-hash`
3. Configure a sharding policy on the input exchange
    - `rabbitmqctl set_policy images-shard "^fluent.modhash$" '{"shards-per-node": 2, "routing-key": "1234"}'`
4. Setup fluentd to use the associated type and bind to a queue named the same as the input exchange name
    - This queue is created 'dynamically' and will not show as a formal queue in the manager, but will deliver events to fluent normally

*Warning*: You will need to run at least N consumers for the N shards created as the plugin does not try to route all shards onto consumers dynamically.

# Contributing to fluent-plugin-amqp <a name="contributing"></a>

- Check out the latest master to make sure the feature hasn't been implemented or the bug hasn't been fixed yet
- Check out the issue tracker to make sure someone already hasn't requested it and/or contributed it
- Fork the project
- Start a feature/bugfix branch
- Commit and push until you are happy with your contribution
- Make sure to add tests for it. This is important so I don't break it in a future version unintentionally.
- Please try not to mess with the Rakefile, version, or history. If you want to have your own version, or is otherwise necessary, that is fine, but please isolate to its own commit so I can cherry-pick around it.

## Releasing

1. Make sure you've got credentials with authorisation to deploy to https://rubygems.org/gems/fluent-plugin-amqp
1. Update the `VERSION` file
1. Run `bundle exec rake release`
1. Run `git push upstream --tags`

# Copyright <a name="copyright"></a>

Copyright (c) 2011 Hiromi Ishii. See LICENSE.txt for
Copyright (c) 2013- github/giraffi. See LICENSE.txt for
further details.


CHANGELOG.md
# 0.14.x

## 0.14.0

* Added
  * Message Headers from source [#62](https://github.com/giraffi/fluent-plugin-amqp/pull/62/) - Thanks @MaheshKReddy

# 0.13.x

## 0.13.0 - Feb 24, 2019

* Feature: TLS Client Certificate and Key are no longer required if TLS is setup - #57
* Added 'auth_mechanism' configuration. - #52

## 0.12.x

## 0.12.0 - Nov 02, 2018

* giraffi/fluent-plugin-amqp#48 - Added support for Multi-Workers

## 0.11.x

## 0.11.0 - March 10, 2017

* Added `content_type` and `content_encoding` configuration for
   the Matcher plugin - #37
* Added header options to set message headers based on the message content on
  the Matcher plugin - #35
* Enabled support for Coveralls code-coverage reporting


# 0.10.x

## 0.10.1 - 2016-11-17

* BugFix: Failed to set ruby_version properly and broke 0.10.0

## 0.10.0 - 2016-11-17

* *Breaking Change* - fluent-amqp-plugin only compatible with fluent >= 0.14.8 and ruby >= 2.1
* Feature: Updated to use new FluentD 0.14 plugin format
* Feature: Support for nanosecond precision

# 0.9.x

## 0.9.3

* BugFix: `:tls => false` hangs connections for some reason #32 HT: @mrkurt

## 0.9.2

* Better error handling in AMQP Matcher to deal with byte range errors, and
any other failure which would prevent buffers from being replayed effectivly.

## 0.9.1

* BugFix: giraffi/fluent-plugin-amqp#25 - Wrapped JSON.dump with simple catch which logs error but does not attempt to fix encoding errors

## 0.9.0

* Added support for Fluent 0.14.x
    * travis-ci build validates plugin against fluent 0.10, 0.12 and 0.14
* Compatibility: Use old json library when building for Ruby 1.9 and Fluent <= 0.12

# 0.8.x

## 0.8.2

* Miscs: *no effect* Remove fixed date from gemspec #22

## 0.8.1

* Fix bug: TypeError with single RabbitMQ. #21

## 0.8.0

* Add automatic failover support for multiple hosts on same cluster. #19

# 0.7.x

## 0.7.1

* Sources can now automatically bind to exchanges [Issue #16](https://github.com/giraffi/fluent-plugin-amqp/issues/16)
* Refactored Docker container to use docker-compose and use multiple containers for testing source and matches concurrently

## 0.7.0

* TLS configuration supported [#15](https://github.com/giraffi/fluent-plugin-amqp/pull/15)
* Docker container created for testing [#13](https://github.com/giraffi/fluent-plugin-amqp/pull/13)

# 0.6.x

## 0.6.1

* Heartbeat added to outbound connections [#11](https://github.com/giraffi/fluent-plugin-amqp/pull/11)

## 0.6.0

*Breaking Change:* Ruby 2.0+ now required

* Updated dependencies for Bunny and Fluent to latest
* Changed default Heartbeat from 0 to 60 seconds

# < 0.5.x

* See commit history for breakdown of changes


test/helper.rb
require 'rubygems'
require 'bundler'
require 'simplecov'
SimpleCov.start

require 'coveralls'
Coveralls.wear!

begin
  Bundler.setup(:default, :development)
rescue Bundler::BundlerError => e
  $stderr.puts e.message
  $stderr.puts "Run `bundle install` to install missing gems"
  exit e.status_code
end

require 'test/unit'
require 'shoulda'

$LOAD_PATH.unshift(File.join(File.dirname(__FILE__), '..', 'lib'))
$LOAD_PATH.unshift(File.dirname(__FILE__))
#require 'fluent-plugin-amqp'


test/plugin/test_in_amqp.rb

require_relative '../helper'
require 'fluent/test'
require 'fluent/test/driver/input'
require 'fluent/test/helpers'
require 'fluent/plugin/in_amqp'

class AMPQInputTest < Test::Unit::TestCase
  include Fluent::Test::Helpers

  CONFIG = %(
    type amqp
    format json
    host amqp.example.com
    port 5672
    vhost /
    user guest
    pass guest
    queue logs
  ).freeze


  TLS_CONFIG = CONFIG + %q(
    tls true
    tls_key "/etc/fluent/ssl/client.key.pem"
    tls_cert "/etc/fluent/ssl/client.crt.pem"
    tls_ca_certificates ["/etc/fluent/ssl/server.cacrt.pem", "/another/ca/cert.file"]
    tls_verify_peer true
  ).freeze

  setup do
    Fluent::Test.setup
    @d = Fluent::Test::Driver::Input.new(Fluent::Plugin::AMQPInput)
  end

  sub_test_case 'configuration' do
    test 'basic configuration' do
      configs = {'basic' => CONFIG}
      configs.merge!('tls' => TLS_CONFIG)

      configs.each_pair { |k, v|
        @d = Fluent::Test::Driver::Input.new(Fluent::Plugin::AMQPInput).configure(v)
        assert_equal "amqp.example.com", @d.instance.host
        assert_equal 5672, @d.instance.port
        assert_equal "guest", @d.instance.user
        assert_equal "/", @d.instance.vhost
      }
    end

    test 'are available with multi worker configuration in default' do
      assert @d.instance.multi_workers_ready?
    end

    test 'array of hosts' do
      conf = CONFIG + %q(
        type amqp
        format json
        hosts ["bob","fred"]
        )

      @d = @d.configure(conf)
      assert_equal ["bob", "fred"], @d.instance.hosts
      assert_equal 5672, @d.instance.port
      assert_equal "guest", @d.instance.user
      assert_equal "/", @d.instance.vhost
    end

    test 'basic TLS without certificates' do
      configs = {'basic' => CONFIG}
      configs.merge!('tls' => CONFIG + %q(tls true))

      configs.each_pair { |k, v|
        @d = Fluent::Test::Driver::Input.new(Fluent::Plugin::AMQPInput).configure(v)
        assert_equal "amqp.example.com", @d.instance.host
        assert_equal 5672, @d.instance.port
        assert_equal "guest", @d.instance.user
        assert_equal "/", @d.instance.vhost
        assert_equal "/", @d.instance.vhost
      }
    end


    test 'non exclusive and multi worker shouldnt change queue name' do
      conf = CONFIG + %(
        type amqp
        format json
        exclusive false
        )

      with_worker_config(workers: 2, worker_id: 0) do
        @d = @d.configure(conf)
        omit("BunnyMock is not avaliable") unless Object.const_defined?("BunnyMock")
        @d.instance.connection = BunnyMock.new

        # Start the driver and wait while it initialises the threads etc
        @d.instance.start
        10.times { sleep 0.05 }
      end
      assert_true @d.instance.connection.queue_exists?("logs")
    end

    test 'exclusive without multiple workers, shouldnt change queue name' do
      conf = CONFIG + %(
        type amqp
        format json
        exclusive true
        )

      @d = @d.configure(conf)
      omit('BunnyMock is not avaliable') unless Object.const_defined?('BunnyMock')
      @d.instance.connection = BunnyMock.new
      # Start the driver and wait while it initialises the threads etc
      @d.instance.start
      10.times { sleep 0.05 }
      assert_true @d.instance.connection.queue_exists?('logs')
    end

    test 'exclusive with multiple workers doesnt update first queue name' do
      conf = CONFIG + %(
        type amqp
        format json
        exclusive true
        )

      with_worker_config(workers: 4, worker_id: 0) do
        @d = @d.configure(conf)
        omit('BunnyMock is not avaliable') unless Object.const_defined?('BunnyMock')
        @d.instance.connection = BunnyMock.new
        # Start the driver and wait while it initialises the threads etc
        @d.instance.start
        10.times { sleep 0.05 }
      end
      assert_true @d.instance.connection.queue_exists?('logs')
    end

    test 'exclusive with multiple workers changes queue name for workers >=1 ' do
      conf = CONFIG + %(
        type amqp
        format json
        exclusive true
        )

      with_worker_config(workers: 4, worker_id: 3) do
        @d = @d.configure(conf)
        omit('BunnyMock is not avaliable') unless Object.const_defined?('BunnyMock')
        @d.instance.connection = BunnyMock.new
        # Start the driver and wait while it initialises the threads etc
        @d.instance.start
        10.times { sleep 0.05 }
      end
      assert_true @d.instance.connection.queue_exists?('logs.3')
    end

    test 'invalid host / queue configuration' do
      assert_raise_message(/'host\(s\)' and 'queue' must be all specified./) do
        @d.configure(%(
          type amqp
          format json
          host bob
          ))
      end
    end
  end

  sub_test_case 'connection handling' do

    test 'queue is created' do
      omit("Need to replace bunny-mock with proper mocking - it cant cope with q.subscribe and sensible replay/verification")
      plugin = @d.configure(CONFIG).instance
      plugin.connection = BunnyMock.new

      # Start the driver and wait while it initialises the threads etc
      plugin.start
      10.times { sleep 0.05 }

      # Should have created the 'logs' queue
      assert_equal true, plugin.connection.queue_exists?('logs')
    end

  end
end


lib/fluent/plugin/in_amqp.rb
require 'time'
require 'fluent/plugin/input'
require 'fluent/plugin/parser'
require 'bunny'

module Fluent::Plugin
  ##
  # AMQPInput to be used as a Fluent SOURCE, reading messages from a RabbitMQ
  # message broker
  class AMQPInput < Input
    Fluent::Plugin.register_input('amqp', self)

    helpers :compat_parameters, :parser

    # Bunny connection handle
    #   - Allows mocking for test purposes
    attr_accessor :connection

    config_param :tag, :string, default: "hunter.amqp"

    config_param :host, :string, default: nil
    config_param :hosts, :array, default: nil
    config_param :user, :string, default: "guest"
    config_param :pass, :string, default: "guest", secret: true
    config_param :vhost, :string, default: "/"
    config_param :port, :integer, default: 5672
    config_param :ssl, :bool, default: false
    config_param :verify_ssl, :bool, default: false
    config_param :heartbeat, :integer, default: 60
    config_param :queue, :string, default: nil
    config_param :durable, :bool, default: false
    config_param :exclusive, :bool, default: false
    config_param :auto_delete, :bool, default: false
    config_param :passive, :bool, default: false
    config_param :payload_format, :string, default: "json"
    config_param :tag_key, :bool, default: false
    config_param :tag_header, :string, default: nil
    config_param :time_header, :string, default: nil
    config_param :tls, :bool, default: false
    config_param :tls_cert, :string, default: nil
    config_param :tls_key, :string, default: nil
    config_param :tls_ca_certificates, :array, default: nil
    config_param :tls_verify_peer, :bool, default: true
    config_param :bind_exchange, :bool, default: false
    config_param :exchange, :string, default: ""
    config_param :routing_key, :string, default: "#"                       # The routing key used to bind queue to exchange - # = matches all, * matches section (tag.*.info)
    config_param :include_headers, :bool, default: false
    config_param :auth_mechanism, :string, default: nil

    def configure(conf)
      conf['format'] ||= conf['payload_format'] # legacy
      compat_parameters_convert(conf, :parser)

      super

      parser_config = conf.elements('parse').first
      if parser_config
        @parser = parser_create(conf: parser_config)
      end

      @conf = conf
      unless (@host || @hosts) && @queue
        raise Fluent::ConfigError, "'host(s)' and 'queue' must be all specified."
      end
    end

    def start
      super
      # Create a new connection, unless its already been provided to us
      @connection = Bunny.new get_connection_options unless @connection
      @connection.start
      @channel = @connection.create_channel

      if @exclusive && fluentd_worker_id > 0
        log.info 'Config requested exclusive queue with multiple workers'
        @queue += ".#{fluentd_worker_id}"
        log.info "Renamed queue name to include worker id: #{@queue}"
      end

      q = @channel.queue(@queue, passive: @passive, durable: @durable,
                       exclusive: @exclusive, auto_delete: @auto_delete)
      if @bind_exchange
        log.info "Binding #{@queue} to #{@exchange}, :routing_key => #{@routing_key}"
        q.bind(exchange=@exchange, routing_key: @routing_key)
      end

      q.subscribe do |delivery, meta, msg|
        log.debug "Recieved message #{msg}"
        log.debug "Recieved message  MetaData #{meta}"
        payload = parse_payload(msg, meta)
        log.debug "Parsed Payload #{payload}"
        router.emit(parse_tag(delivery, meta), parse_time(meta), payload)
      end
    end # AMQPInput#run

    def shutdown
      log.info "Closing connection"
      @connection.stop
      super
    end

    def multi_workers_ready?
      true
    end

    private
    def parse_payload(msg, meta)
      parsed = nil
      if @parser
        @parser.parse msg do |_, payload|
          if payload.nil?
            log.warn "failed to parse #{msg}"
            parsed = { 'message' => msg }
          else
            parsed = payload
          end
        end
      else
        parsed = { 'message' => msg }
      end
      if @include_headers
        log.debug 'Adding headers'
        { 'headers' => meta[:headers] }.merge(parsed)
      else
        parsed
      end
    end

    def parse_tag( delivery, meta )
      if @tag_key && delivery.routing_key != ''
        delivery.routing_key
      elsif @tag_header && meta[:headers][@tag_header]
        meta[:headers][@tag_header]
      else
        @tag
      end
    end

    def parse_time( meta )
      if @time_header && meta[:headers][@time_header]
        Fluent::EventTime.from_time(Time.parse( meta[:headers][@time_header] ))
      else
        Fluent::Engine.now
      end
    end

    def get_connection_options()
      hosts = @hosts ||= Array.new(1, @host)
      opts = {
        hosts: hosts, port: @port, vhost: @vhost,
        pass: @pass, user: @user, ssl: @ssl,
        verify_ssl: @verify_ssl, heartbeat: @heartbeat,
        tls: @tls,
        verify_peer: @tls_verify_peer,
        auth_mechanism: @auth_mechanism
      }
      # Include additional optional TLS configurations
      opts[:tls_key] = @tls_key if @tls_key
      opts[:tls_cert] = @tls_cert if @tls_cert
      opts[:tls_ca_certificates] = @tls_ca_certificates if @tls_ca_certificates
      return opts
    end

  end # class AMQPInput

end # module Fluent::Plugin


lib/fluent/plugin/out_amqp.rb
require 'json'
require 'fluent/plugin/output'
require "bunny"

module Fluent::Plugin
  ##
  # AMQPOutput to be used as a Fluent MATCHER, sending messages to a RabbitMQ
  # messaging broker
  class AMQPOutput < Output
    Fluent::Plugin.register_output("amqp", self)

    helpers :compat_parameters

    DEFAULT_BUFFER_TYPE = "memory"

    attr_accessor :connection

    #Attribute readers to support testing
    attr_reader :exch
    attr_reader :channel


    config_param :host, :string, default: nil
    config_param :hosts, :array, default: nil
    config_param :user, :string, default: "guest"
    config_param :pass, :string, default: "guest", secret: true
    config_param :vhost, :string, default: "/"
    config_param :port, :integer, default: 5672
    config_param :ssl, :bool, default: false
    config_param :verify_ssl, :bool, default: false
    config_param :heartbeat, :integer, default: 60
    config_param :exchange, :string, default: ""
    config_param :exchange_type, :string, default: "direct"
    config_param :passive, :bool, default: false
    config_param :durable, :bool, default: false
    config_param :auto_delete, :bool, default: false
    config_param :key, :string, default: nil
    config_param :persistent, :bool, default: false
    config_param :tag_key, :bool, default: false
    config_param :tag_header, :string, default: nil
    config_param :time_header, :string, default: nil
    config_param :tls, :bool, default: false
    config_param :tls_cert, :string, default: nil
    config_param :tls_key, :string, default: nil
    config_param :tls_ca_certificates, :array, default: nil
    config_param :tls_verify_peer, :bool, default: true
    config_param :content_type, :string, default: "application/octet"
    config_param :content_encoding, :string, default: nil
    config_param :auth_mechanism, :string, default: nil

    config_section :header do
      config_set_default :@type, DEFAULT_BUFFER_TYPE
    end

    config_section :buffer do
      config_set_default :@type, DEFAULT_BUFFER_TYPE
    end


    class HeaderElement
      include Fluent::Configurable

      config_param :name, :string
      config_param :default, :string, default: nil
      config_param :source, default: nil  do |val|
             if val.start_with?('[')
              JSON.load(val)
             else
               val.split('.')
            end
         end

      # Extract a header and value from the input data
      # returning nil if value cannot be derived
      def getValue(data)
        val  = getNestedValue(data, @source ) if @source
        val ||= @default if @default
        val
      end

      def getNestedValue(data, path)
        temp_data = data
        temp_path = path.dup
        until temp_data.nil? or temp_path.empty?
          temp_data = temp_data[temp_path.shift]
        end
        temp_data
      end
    end

    def configure(conf)
      compat_parameters_convert(conf, :buffer)
      super
      @conf = conf

      # Extract the header configuration into a collection
      @headers = conf.elements.select {|e|
        e.name == 'header'
      }.map {|e|
        he = HeaderElement.new
        he.configure(e)
        unless he.source || he.default
            raise Fluent::ConfigError, "At least 'default' or 'source' must must be defined in a header configuration section."
        end
        he
      }

      unless @host || @hosts
        raise Fluent::ConfigError, "'host' or 'hosts' must be specified."
      end
      unless @key || @tag_key
        raise Fluent::ConfigError, "Either 'key' or 'tag_key' must be set."
      end
    end

    def start
      super
      begin
        log.info "Connecting to RabbitMQ..."
        @connection = Bunny.new(get_connection_options) unless @connection
        @connection.start
      rescue Bunny::TCPConnectionFailed => e
        log.error "Connection to #{@host} failed"
      rescue Bunny::PossibleAuthenticationFailureError => e
        log.error "Could not authenticate as #{@user}"
      end

      log.info "Creating new exchange #{@exchange}"
      @channel = @connection.create_channel
      @exch = @channel.exchange(@exchange, type: @exchange_type.intern,
                              passive: @passive, durable: @durable,
                              auto_delete: @auto_delete)
    end

    def shutdown
      @connection.stop
      super
    end

    def multi_workers_ready?
      true
    end

    def formatted_to_msgpack_binary
      true
    end

    def format(tag, time, record)
      [tag, time, record].to_msgpack
    end

    def write(chunk)
      begin
        chunk.msgpack_each do |(tag, time, data)|
          begin
            msg_headers = headers(tag,time,data)

            begin
              data = JSON.dump( data ) unless data.is_a?( String )
            rescue JSON::GeneratorError => e
              log.warn "Failure converting data object to json string: #{e.message} - sending as raw object"
              # Debug only - otherwise we may pollute the fluent logs with unparseable events and loop
              log.debug "JSON.dump failure converting [#{data}]"
            end

            log.debug "Sending message #{data}, :key => #{routing_key( tag)} :headers => #{headers(tag,time,data)}"
            @exch.publish(
              data,
              key: routing_key( tag ),
              persistent: @persistent,
              headers: msg_headers,
              content_type: @content_type,
              content_encoding: @content_encoding)

  # :nocov:
  #  Hard to throw StandardError through test code
          rescue StandardError => e
            # This protects against invalid byteranges and other errors at a per-message level
            log.error "Unexpected error during message publishing: #{e.message}"
            log.debug "Failure in publishing message [#{data}]"
          end
        end
      rescue MessagePack::MalformedFormatError => e
        # This has been observed when a server has filled the partition containing
        # the buffer files, and during replay the chunks were malformed
        log.error "Malformed msgpack in chunk - Did your server run out of space during buffering? #{e.message}"
      rescue StandardError => e
        # Just in case theres any other errors during chunk loading.
        log.error "Unexpected error during message publishing: #{e.message}"
      end
      # :nocov:
    end


    def routing_key( tag )
      if @tag_key
        tag
      else
        @key
      end
    end

    def headers( tag, time, data )
      h = {}

      log.debug "Processing Headers: #{@headers}"
      # A little messy this...
      # Trying to allow for header overrides where a header defined
      # earlier will be used if a later header is returning nil (ie not found and no default)
      h = Hash[ @headers
                  .collect{|v| [v.name, v.getValue(data) ]}
                  .delete_if{|x| x.last.nil?}
          ]

      h[@tag_header] = tag if @tag_header
      h[@time_header] = Time.at(time).utc.to_s if @time_header

      h
    end

    def get_connection_options()
      hosts = @hosts ||= Array.new(1, @host)
      opts = {
        hosts: hosts, port: @port, vhost: @vhost,
        pass: @pass, user: @user, ssl: @ssl,
        verify_ssl: @verify_ssl, heartbeat: @heartbeat,
        tls: @tls || nil,
        tls_cert: @tls_cert,
        tls_key: @tls_key,
        verify_peer: @tls_verify_peer,
        auth_mechanism: @auth_mechanism
      }
      opts[:tls_ca_certificates] = @tls_ca_certificates if @tls_ca_certificates
      return opts
    end
  end
end
